File: contents/forward_euler_method/code/java/Euler.java
public class ForwardEuler {
  
  public static double[] append(double[] oldArray, double newItem) {
    double[] newArray = new double[oldArray.length + 1];

    for(int i = 0; i < oldArray.length; i++) {
        newArray[i] = oldArray[i];
    }
    newArray[newArray.length - 1] = newItem;

    return newArray;
}
  
  public static double[] solveEuler(double timeStep, int n) {
    double[] result = {1};

    for(int i = 0; i < n; i++) {
      result = append(result, result[i] - 3 * result[i] * timeStep);
    }

    return result;
  }
  
  public static boolean checkResult(double[] result, double threshold, double timeStep) {
    boolean isApprox = true;

    for(int i = 0; i < result.length; i++) {
        double solution = Math.pow(Math.E, -3 * i * timeStep);
      
        if(Math.abs(result[i] - solution) > threshold) {
            isApprox = false;
        }
    }

    return isApprox;
}

  public static void main(String[] args) {
    double timeStep = 0.01;
    int n = 100;
    double threshold = 0.01;
    
    double[] result = solveEuler(timeStep, n);
    boolean isApprox = checkResult(result, threshold, timeStep);

    if(isApprox) {
        System.out.print("All values within threshold");
    } else {
        System.out.print("Value(s) not in threshold");
    } 
  }
}

File: contents/forward_euler_method/code/swift/euler.swift
import Foundation

func solveEuler(timeStep: Double, n: Int) -> [Double] {
    var result : [Double] = [1]

    for i in 1...n {
        result.append(result[i - 1] - 3 * result[i - 1] * timeStep)
    }
    
    return result
}

func checkResult(result: [Double], threshold: Double, timeStep: Double) -> Bool {
    var isApprox = true
    
    for i in 0..<result.count {
        let solution = exp(-3 * Double(i) * timeStep)
        if abs(result[i] - solution) > threshold {
            print(result[i], solution)
            isApprox = false
        }
    }
    
    return isApprox
}

func main() {
    let timeStep = 0.01
    let n = 100
    let threshold = 0.01

    let result = solveEuler(timeStep: timeStep, n: n)
    let isApprox = checkResult(result: result, threshold: threshold, timeStep: timeStep)

    if isApprox {
        print("All values within threshold")
    } else {
        print("Value(s) not in threshold")
    }
}

main()
